### ISSUE TITLE
Linux with docker built Kivy Wheel - will not use sdl2

### ISSUE THREAD

**Issue by cp-jpatterson** [kivy/kivy#5803]
### Versions

* Python: 2.7.13
* OS: Centos 7
* Kivy: 1.10.0 (and 1.10.1.dev0)
* Kivy installation method: pip wheel kivy -> pip (install wheel, on target system)

### Description

To avoid rebuilding kivy on our target systems, we want to be able to deliver a kivy wheel that we instead deliver (the target system doesn't have direct network connectivity or any way to build).

### Code and Logs
The wheel build system has:
        SDL2-devel
        SDL2_mixer-devel
        SDL2_ttf-devel
        SDL2_image-devel
        SDL2_gfx-devel
and 
        mesa-libGL
        mesa-libGL-devel
        mesa-libGLES
        mesa-libGLES-devel
        mesa-libEGL
        mesa-libEGL-devel
and
        khrplatform-devel

The target system has:
        khrplatform
        mesa-libGL
        mesa-libGLES
        mesa-libEGL
        SDL2
        SDL2_mixer
        SDL2_ttf
        SDL2_image
        SDL2_gfx

Are there other dependencies in either the build or target that sdl2 support would depend on?

**Comment by tshirtman** [comment]
it could help to have the build log in the docker, but maybe try setting `USE_SDL2` in the env before the build.

**Comment by cp-jpatterson** [comment]
Thanks @tshirtman - There's precious little provided in the output from pip :(  I'll try setting USE_SDL2

**Comment by tshirtman** [comment]
Maybe before trying to use pip, try building with `python setup.py bdist_wheel` or even `python setup.py build_ext --inplace` to see what happens at build time.

**Comment by dolang** [comment]
Are the wheel build system and the target system actually the same? Could you add some more information for:
* actual error you're seeing
* build procedure
* glibc versions if they're different (+gcc version)
* any other version differences

Or maybe the SDL2 libraries aren't loaded, and some `LD_PRELOAD` or [`SDL_DYNAMIC_API`](https://hg.libsdl.org/SDL/file/default/docs/README-dynapi.md#l64) trickery could help?

**Comment by cp-jpatterson** [comment]
Hi thanks @dolang 

The actual error I'm seeing is it's picking up pygame instead of SDL2
The build procedure is simply to try pip wheel kivy
I've used the standard CentOS glibc, etc..

It's certainly possible that some trickery is necessary.

I assume there's some set of flags, libraries that will produce a wheel on linux for kivy that will utilize SDL2.....is that a valid assumption?   Has that assumption been tested successfully?

**Comment by cp-jpatterson** [comment]
BTW...I'm working on building with "python setup.py bdist_wheel" instead of "pip wheel kivy" to see if I can get any more useful output that way.

export USE_SDL=1  didn't seem to help...and I know that the build somewhere was picking it up because I first did:
export USE_SDL=True
lol - which caused the pip wheel kivy command to fail :)

**Comment by dolang** [comment]
`export USE_SDL2=1` would actually be the correct way to enable that.

It's good that both build and target system are CentOS 7, but it might still matter that you're using Ubuntu instead. However, it's probably easier if you join the [#kivy IRC channel](https://kivy.org/#support) to discuss this.

**Comment by cp-jpatterson** [comment]
Sorry right, I am doing 
export USE_SDL2=1

Typing faster than my coding angels can fly......sorry about that

**Comment by cp-jpatterson** [comment]
OK, I was finally able to make a bit of progress on @tshirtman 's suggestion:

```
Using distutils
Environ change use_sdl2 -> True

Detected Cython version 0.25.2
Using this graphics system: OpenGL
WARNING: A problem occurred while running pkg-config --libs --cflags gstreamer-1.0 (code 1)

Package gstreamer-1.0 was not found in the pkg-config search path.
Perhaps you should add the directory containing `gstreamer-1.0.pc'
to the PKG_CONFIG_PATH environment variable
No package 'gstreamer-1.0' found


WARNING: A problem occurred while running pkg-config --libs --cflags sdl2 SDL2_ttf SDL2_image SDL2_mixer (code 1)

Package sdl2 was not found in the pkg-config search path.
Perhaps you should add the directory containing `sdl2.pc'
to the PKG_CONFIG_PATH environment variable
No package 'sdl2' found
Package SDL2_ttf was not found in the pkg-config search path.
Perhaps you should add the directory containing `SDL2_ttf.pc'
to the PKG_CONFIG_PATH environment variable
No package 'SDL2_ttf' found
Package SDL2_image was not found in the pkg-config search path.
Perhaps you should add the directory containing `SDL2_image.pc'
to the PKG_CONFIG_PATH environment variable
No package 'SDL2_image' found
Package SDL2_mixer was not found in the pkg-config search path.
Perhaps you should add the directory containing `SDL2_mixer.pc'
to the PKG_CONFIG_PATH environment variable
No package 'SDL2_mixer' found


SDL2: missing sub library SDL
SDL2: missing sub library SDL_mixer
SDL2: missing sub library SDL_ttf
SDL2: missing sub library SDL_image
```

**Comment by dolang** [comment]
Found something, but don't know anymore what exactly I did with it. But you can try:
```
pkg-config --list-all
```
and see if it finds SDL2. Also there's the environment variable `PKG_CONFIG_PATH`. There should be some `.pc` file for your SDL2 library somewhere, and adding the directory (I think) containing that to the env variable should help.

**Comment by matham** [comment]
Please take a look at our .travis file. That does a complete self contained
build and wheel building on centos 7 docker.

**Comment by dolang** [comment]
> 1. Does doing the CFLAGS= mess something else up? (I suspect maybe yes) Is there a better suggested way to get the -D's necessary to get around the problem? I'm not fond of the idea of downgrading the SDL2 libraries, but will if that's the only way out.

Your workaround should be unproblematic. It was fixed in #5459 by just removing the offending flags, because no one is using a setup like that anyway, I guess.

And it's really only that minor source incompatibility that's breaking the build.

> 2. If CFLAGS isn't the issue, then I'm still a bit stumped....I'm having problems in the environment with my test build script that runs python setup.py bdist_wheel, setup.py starts, but then fails to find bdist_wheel - I have done pip install wheel already, so I know that's not the issue there.
    Another point to note is that even with CFLAGS defined, the python setup.py bdist_wheel build was still finding the SDL2 libraries installed. Not surprising, but it's something to rule out.

Do you have both `setuptools` and `wheel`? Maybe you need to convince Kivy to use `setuptools` with `KIVY_USE_SETUPTOOLS=1`.

It supports both `distutils` by itself and `setuptools`, but I don't know if `distutils` has the necessary hooks to create wheels.

**Comment by cp-jpatterson** [comment]
Well - ok, ..... that build actually produced a wheel that will use sdl2, I stand corrected.  I just tested it with a new run to get that debug output....not sure what I did to foobar my previous test.

Thanks everyone for your help with this

**Comment by dolang** [comment]
I haven't been involved long enough to know about all the details, but from personal experience: you **should not** set `USE_X11`, and you should not **need** to specify `USE_SDL2` on Linux, as that is set automatically when pkgconfig works right and all four SDL2 libraries are present.

If you're doing some builds on Windows too, then `USE_SDL2` is probably necessary.

**Comment by cp-jpatterson** [comment]
Yeah - it may have been that the earlier build I tested had USE_X11 set?  I took that out before I rebuilt.

Thanks!
